_posts/2024-07-28-Uncovering-Code-Coverage.md

```diff
@@ -0,0 +1,229 @@
+Let's get started!
+
+<!-- excerpt end -->
+
+## What is Code Coverage?
+
+{% include figure.html image="images/simple-example-of-coverage-for-functions.png" width="600px" caption="Code which is tested is considered 'covered'." %}
+
+> "With tests, we can change the behavior of our code quickly and verifiably. Without them, we really don't know if our code is getting better or worse." ([Michael Feathers, Working Effectively with Legacy Code](https://www.goodreads.com/quotes/718460-code-without-tests-is-bad-code-it-doesn-t-matter-how))
+
+Code coverage, also known as test coverage, refers to the percentage of your code that is executed when your tests run ([Wikipedia: Code coverage](https://en.wikipedia.org/wiki/Code_coverage)).
+Software developers use code coverage as a way to understand areas which may need tests.
+By understanding these metrics, you can get a clearer picture of your code's reliability.
+
+## Types of Coverage
+
+```python
+# example line
+print("A line which could be tested for coverage.")
+
+# example statement (potentially multi-line)
+statement = [
+    1,
+    2,
+    3,
+]
+
+# example function
+def example():
+    return "An example function."
+
+# example branching
+if 1 in statement:
+    return True
+else:
+    return False
+```
+
+Line coverage, statement coverage, function coverage, and branch coverage common metrics used to measure how thoroughly code is tested, but they each focus on different aspects.
+
+- **Line coverage** tracks the execution of individual lines of code, indicating whether each line has been run during testing.
+- **Statement coverage** is similar but considers whether each statement (a single unit of execution) within the code has been executed, regardless of the number of lines.
+- **Function coverage** checks whether each function or method in the code has been called during the tests, ensuring that all parts of the code's interface are exercised.
+- **Branch coverage**, on the other hand, goes deeper by verifying whether every possible path through control structures (like if-else statements) has been followed, ensuring that all potential outcomes and decision points are tested. 
+
+## Benefits of Test Coverage
+
+> "If you are testing thoughtfully and well, I would expect a coverage percentage in the upper 80s or 90s. I would be suspicious of anything like 100% - it would smell of someone writing tests to make the coverage numbers happy, but not thinking about what they are doing." ([Martin Fowler, Test Coverage](https://martinfowler.com/bliki/TestCoverage.html))
+
+High test coverage ensures your code is reliable and less prone to bugs. 
+It helps you identify untested parts of your codebase, facilitating better maintenance and encouraging a culture of quality.
+This is balanced by understanding that 100% coverage is often very challenging to achieve and could be harmful in practice.
```

Comment:
Maybe the keyword "anti-pattern" could be thrown in somewhere in the post. I've seen a lot of people say 100% coverage is an anti-pattern that just wastes time, but not sure how much of a CS consensus that is.

Reply:
Thanks, I think that's appropriate, I'll work this in!